File: 백준/Silver/2178. 미로 탐색/미로 탐색.js
const input = require('fs').readFileSync(0).toString().trim().split('\n');
let idx = 0;
const [N, M] = input[idx++].split(' ').map(Number);
const maze = [];
for (let i = 0; i < N; i++) {
    maze.push(input[idx++].trim().split('').map(Number));
}

const visited = Array.from({length: N}, () => new Array(M).fill(false));
const queue = [[0, 0, 1]];
visited[0][0] = true;
let front = 0;

while (front < queue.length) {
    const [row, col, dist] = queue[front++];
    
    if (row === N-1 && col === M-1) {
        console.log(dist);
        break;
    }
    
    for (const [dr, dc] of [[-1,0],[1,0],[0,-1],[0,1]]) {
        const nr = row + dr;
        const nc = col + dc;
        
        if ((nr >= 0 && nr < N) && (nc >= 0 && nc < M) && maze[nr][nc] === 1 && !visited[nr][nc]) {
            visited[nr][nc] = true;
            queue.push([nr, nc, dist + 1]);
        }
    }
}

File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
from collections import deque
import sys
input = sys.stdin.readline

N, M = map(int, input().split())
maze = [list(map(int, input().strip())) for _ in range(N)]
visited = [[False] * M for _ in range(N)]

queue = deque([(0, 0, 1)])
visited[0][0] = True

while queue:
    row, col, dist = queue.popleft()
    
    if row == N-1 and col == M-1:
        print(dist)
        break
    
    for dr, dc in [(-1,0),(1,0),(0,-1),(0,1)]:
        nr, nc = row + dr, col + dc
        if 0 <= nr < N and 0 <= nc < M and maze[nr][nc] == 1 and not visited[nr][nc]:
            visited[nr][nc] = True
            queue.append((nr, nc, dist + 1))